independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/RunnerClassLoader.java

```diff
@@ -73,33 +71,35 @@
             if (loaded != null) {
                 return loaded;
             }
-            ClassLoadingResource[] resources;
+            final ClassLoadingResource[] resources;
             if (packageName == null) {
                 resources = resourceDirectoryMap.get("");
             } else {
-                String dirName = packageName.replace(".", "/");
+                String dirName = packageName.replace('.', '/');
                 resources = resourceDirectoryMap.get(dirName);
             }
             if (resources != null) {
-                String classResource = name.replace(".", "/") + ".class";
+                String classResource = name.replace('.', '/') + ".class";
                 for (ClassLoadingResource resource : resources) {
                     byte[] data = resource.getResourceData(classResource);
                     if (data == null) {
                         continue;
                     }
                     definePackage(packageName, resources);
                     return defineClass(name, data, 0, data.length,
-                            protectionDomains.computeIfAbsent(resource, new Function<ClassLoadingResource, ProtectionDomain>() {
-                                @Override
-                                public ProtectionDomain apply(ClassLoadingResource ce) {
-                                    return ce.getProtectionDomain(RunnerClassLoader.this);
-                                }
-                            }));
+                            getProtectionDomain(resource));
                 }
             }
         }
         return getParent().loadClass(name);
+    }
+
+    private ProtectionDomain getProtectionDomain(ClassLoadingResource resource) {
+        return protectionDomains.computeIfAbsent(resource, this::extractProtectionDomainFromClassLoadingResource);
```

Comment:
Are we sure about using a method reference here?

Reply:
why not? It's typically preferreable over a lambda - although in this case (since my other changes made it stateless) it should be comparable.

I generally highly prefer method references, as there is no possible mistake about the lambda possibly becoming a stateful lambda, which requires extra memory, a proxy, and extra overhead to track the proxy. A method handle will never unintentionally capture state, so it helps to fight such hidden costs.

Reply:
It doesn't capture state, but it's still implemented via `invokedynamic`, isn't it?

Reply:
I know it should not make any real world difference and the cost will probably be dominated by the slowness of `computeIfAbsent`, just trying to cover all bases because we are talking about a core component :)

Reply:
yes but `invokedynamic` isn't a performance problem? Capturing state lambdas and anonymous Functiona classes are far worse

Reply:
BTW I thought this was straight forward, but if there's concerns I don't mind removing this specific commit. It's not that important :)

I do care more about some of the other commits in this PR though, so I'd prefer removing this change if it expedites the others.

Reply:
Right, if it were me, I'd remove it, but I'm not hell bent on it

Reply:
removed

Reply:
Is there a reason why you don't apply your trick of trying to get the value first before calling `computeIfAbsent`? Or is it a case where you have a good change for it to not be in the map?

You told me about it a long time ago and it had quite a few benefits in HV.

Reply:
yea good idea, we could have done that as well I guess. I didn't jump to it as this whole block doesn't seem highly concurrent - call me conservative, but @geoand still got scared :)

Reply:
BTW I also wonder why do we care for protection domains... return a singleton and disable all this stuff :dagger: 

But I don't know the consequences well enough.

Reply:
In this specific case, that should work as there is ever only one `RunnerClassLoader`, so the `JarResource` could indeed be passed the ClassLoader that it helps "populate" (perhaps by adding an `init` method to `ClassLoadingResource`) and then this block would just class `getProtectionDomain` which would just return the singleton.

Reply:
I can make that change when this lands

Reply:
Here it is: https://github.com/quarkusio/quarkus/pull/13515